actors/miner/src/lib.rs

```diff
@@ -4652,6 +4875,122 @@ fn validate_precommits(
     Ok((batch.gen(), verify_infos))
 }
 
+fn validate_ni_sectors(
+    rt: &impl Runtime,
+    sectors: &[SectorNIActivationInfo],
+    seal_proof_type: RegisteredSealProof,
+    all_or_nothing: bool,
+) -> Result<(BatchReturn, Vec<SectorSealProofInput>, BitField), ActorError> {
+    let receiver = rt.message().receiver();
+    let miner_id = receiver.id().unwrap();
+    let curr_epoch = rt.curr_epoch();
+    let activation_epoch = curr_epoch;
+    let challenge_earliest = curr_epoch - rt.policy().max_prove_commit_ni_randomness_lookback;
+    let unsealed_cid = CompactCommD::empty().get_cid(seal_proof_type).unwrap();
+    let entropy = serialize(&receiver, "address for get verify info")?;
+
+    if sectors.is_empty() {
+        return Ok((BatchReturn::empty(), vec![], BitField::new()));
+    }
+    let mut batch = BatchReturnGen::new(sectors.len());
+
+    let mut verify_infos = vec![];
+    let mut sector_numbers = BitField::new();
+    for (i, sector) in sectors.iter().enumerate() {
+        let mut fail_validation = false;
+
+        let set = sector_numbers.get(sector.sector_number);
+        if set {
+            return Err(actor_error!(
+                illegal_argument,
+                "duplicate sector number {}",
+                sector.sector_number
+            ));
+        }
+
+        if sector.sector_number > MAX_SECTOR_NUMBER {
+            warn!("sector number {} out of range 0..(2^63-1)", sector.sector_number);
+            fail_validation = true;
+        }
+
+        sector_numbers.set(sector.sector_number);
+
+        if let Err(err) = validate_expiration(
+            rt.policy(),
+            curr_epoch,
+            activation_epoch,
+            sector.expiration,
+            seal_proof_type,
+        ) {
+            warn!("invalid expiration: {}", err);
+            fail_validation = true;
+        }
+
+        if sector.sealer_id != miner_id {
+            warn!("sealer must be the same as the receiver actor for all sectors");
+            fail_validation = true;
+        }
+
+        if sector.sector_number != sector.sealing_number {
+            warn!("sealing number must be same as sector number for all sectors");
+            fail_validation = true;
+        }
+
+        if !is_sealed_sector(&sector.sealed_cid) {
+            warn!("sealed CID had wrong prefix");
+            fail_validation = true;
+        }
+
+        if sector.seal_rand_epoch >= curr_epoch {
+            warn!(
+                "seal challenge epoch {} must be before now {}",
+                sector.seal_rand_epoch, curr_epoch
+            );
+            fail_validation = true;
```

Comment:
```suggestion
            return Err(actor_error!(
                illegal_argument,
                "seal challenge epoch {} must be before now {}",
                sector.seal_rand_epoch, curr_epoch
            ));
```

just fail the whole lot in this case I think